[WIP] VASP/Phonopy Phonon Workflow

JaGeo opened the pull request:
## Summary

This is a work in progress implementation of a Phonon Workflow with VASP and Phonopy. It is based on the work and experiences of @ab5424 in his Master's thesis (for atomate).

I'm still a long way from finishing the workflow, but I believe I've identified the most critical classes that need to be developed and have an idea regarding data transfer.

Of course, feedback is welcome now, but I believe it requires some more changes from my side before someone of you invests too much time.

JaGeo commented:
As I know that @gpetretto and @davidwaroquiers are also working on transferring workflows from abiflows to atomate2, there might be already some ideas about a Document for PhononData. If so, I would be happy to hear about it.

utf commented:
Thanks very much for this Janine!

I'll aim to take a proper look later this week. 

Would you mind installing the pre-commit? This will automatically raise (and most likely fix) the linting errors. To install you can run the following in the root directory:

```
pip install pre-commit
pre-commit install
```

For this first time, you can run the pre-commit on all files:
```
pre-commit run --all
```

Afterwards it will get run automatically whenever you commit something.

JaGeo commented:
Thanks. Will do. As I said, it really still needs some days of work and testing but it probably helps to get some suggestions with regard to general design now.

gpetretto commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,163 @@
+        jobs.append(vasp_displacement_calcs)
+
+        # Computation of BORN charges
+        if self.born_maker is None:
```
I have two comments about this point: 
1) Maybe you already planned to update this later, but for how it is written now the BECs are always calculated. However there are good reasons to skip this step (e.g. metals). It would be good that `None` means that BECs are not calculated.
2) with the `born_maker: BaseVaspMaker = field(default_factory=StaticSetGenerator)` it will enter this `if` only if born_maker is explicitly set to `None`. Also the default of the of born_maker will not include lepsilon. An option would be to define the born_maker:
`born_maker: BaseVaspMaker = field(default_factory=lambda: StaticSetGenerator(lepsilon=True))`

JaGeo replied:
Thank you for pointing this out. I will add this option by using `born_maker: BaseVaspMaker |None = field(default_factory=lambda: StaticSetGenerator(lepsilon=True))` similar to what is done for the `bulk_relax_maker`. I was actually thinking about doing this but somehow did not completely implement this option ...

JaGeo commented:
As mentioned, I have added one of the suggestions by Guido and the pre-commit flags. I am happy to hear more. Otherwise, I hope that I can refine the workflow in the coming weeks. 
In parts of the workflow, I am using directories to access the files to create the output data. I guess it would be much cleaner if I would transfer forces and born data as outputs. This would be something, where I would appreciate an opinion.

JaGeo commented:
Switched to transfer via outputs for now. I think this should simplify extending the workflow to other codes as well.

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,409 @@
+    return displacements
+
+
+@job(output_schema=PhononBSDOSDoc)
```
I think phonon band structures and density of states should be stored in the data store (for large objects). You can do this automatically using:

```suggestion
from pymatgen.phonon.bandstructure import PhononBandStructure
from pymatgen.phonon.dos import PhononDos

@job(output_schema=PhononBSDOSDoc, data=[PhononDos, PhononBandStructure])
```

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,409 @@
+    #     print(born)
+    if born is not None:
+        if not conventional:
+            primitive = [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]
```
I think phonopy now has tools to determine the primitive matrix automatically?

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,409 @@
+@job
+def run_phonon_displacements(
+    displacements,
+    phonon_maker: BaseVaspMaker = None,
```
I think it would be useful to include some metadata with these displacements. The metadata will automatically get added to the VASP task document. The metadata I have in mind is the index of the displacement, the original (non-supercell) structure, and the bulk supercell structure. This will help reconstructing the phonon documents later a bit easier!

You can see how I did this in the electron phonon workflow. Specifically, this function: https://github.com/materialsproject/atomate2/blob/9a7c3ccdadbbc617eb164e6ae1ff801aae2d1c31/src/atomate2/vasp/jobs/elph.py#L121

And the metadata is included using code such as this:

```python
        # write details of the electron phonon temperature and structure elph_info.json
        # file. this file will automatically get added to the task document and allow
        # the elph builder to reconstruct the elph document. note the ":" is
        # automatically converted to a "." in the filename.
        info = {
            "displacement_number": i,
            "original_structure": original_structure,
            "supercell_structure": supercell_structure,
        }
        elph_job.update_maker_kwargs(
            {"_set": {"write_additional_data->elph_info:json": info}}, dict_mod=True
        )
```

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,409 @@
+def get_phonon_object(
+    displacement, min_length, structure, sym_reduce, symprec, conventional
+):
+    transformation = CubicSupercellTransformation(min_length=min_length)
```
Rather than having this here (where `CubicSupercellTransformation` will get called multiple times at different parts of the workflow), could it be refactored so that the supercell matrix is calculated once at the beginning of the flow and then passed directly to the various other jobs throughout.

JaGeo replied:
Yes, agreed. I will indeed change this.

utf reviewed (commented):
Hi @JaGeo, apologies its taken some time to look at this in detail.

Overall it looks great! This is exactly along the lines of what I envisioned. I definitely think using the outputs directly (rather than files) is a better approach. I don't think the amount of data transfer will be too bad.

I have two main comments:
1. We need to be careful about relying on objects inside of Flow Maker make functions. Any function that actually requires the objects should be wrapped in a job.
2. I think it would be useful to refactor the `generate_frequencies_eigenvectors` so that most of the work is done inside the pydantic model.

I've gone into a lot more detail in the code review.

If anything isn't clear, I'd be happy to discuss on here/email/zoom. Thanks again for this. I'm very excited to have a phonon workflow in atomate2.

JaGeo commented:
Thank you very much for the detailed review. I will try to improve the code according to these suggestions. I hope I can manage to find time in the next 1-2 weeks :).

JaGeo commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,226 @@
+    def make(
+        self,
+        structure: Structure,
+        prev_vasp_dir: str | Path | None = None,
```
I agree with you. I am also not completely happy with the CubicSupercellTransformation a the moment. It prefers similar lattice parameters over 90 degree angles and I am not really sure this is optimal for these kind of phonon computations.

utf replied:
ASE has a better scheme for generating a cubic supercell. See the function here https://wiki.fysik.dtu.dk/ase/ase/build/tools.html#ase.build.find_optimal_cell_shape

And the explanation here: https://wiki.fysik.dtu.dk/ase/tutorials/defects/defects.html#algorithm-for-finding-optimal-supercell-shapes

Perhaps we could use these ase functions instead?

JaGeo replied:
I think I would need to test this in much more detail but it does look very useful especially with access to the optimality measure.

JaGeo commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,226 @@
+    symprec: float = SETTINGS.SYMPREC
+    displacement: float = 0.01
+    min_length: float = 20.0
+    conventional: bool = False
```
Agreed but then we should probably also make the computation of the band structure optional?

utf replied:
Potentially? Although it might be possible to use [the AUTO option for PRIMITIVE_AXIS](https://phonopy.github.io/phonopy/setting-tags.html#primitive-axes-or-primitive-axis) to automatically align the band structure whatever input structure you use?

JaGeo replied:
Yes, true.

JaGeo replied:
Yes, I know this option. Need to check.

JaGeo replied:
I would like to be able to use the kpath options from pymatgen. I am not sure phonopy uses the same convention for the primitive cell.

JaGeo commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,226 @@
+
+        # Computation of BORN charges
+        if self.born_maker is not None:
+            born_job = StaticMaker(input_set_generator=self.born_maker).make(
```
Thanks! Understood!

utf commented:
The changes you've made already look great! Happy to help diagnose any bugs should they come up.

JaGeo commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,418 @@
+    # I don't see it
+    # I think I am missing something in the code
+    phonon.save("phonopy.yaml")
+    phonon = load("phonopy.yaml")
```
@utf : I currently have a problem here. I can only get correct bandstructures with NAC correction when I save the phonon object and then reload. There should be another way but I currently don't find the bug. If you have any idea, I am happy to hear.

JaGeo replied:
I am attaching minimal example where it does not happen. I can currently not reproduce this bug outside of the workflow. 
[minimal_phonopy.zip](https://github.com/materialsproject/atomate2/files/8879780/minimal_phonopy.zip)

utf replied:
I'm afraid I can't see what is going wrong!

Can you confirm that you're using the same phonopy version outside of the workflow?

JaGeo replied:
I will confirm this and check again but thanks already for having a look!

JaGeo replied:
I have upgraded to the latest phonopy version. It is still happening. I think I will continue with this file creation part for now and try to finalize the document part first. 
![phonon_band_structure-01](https://user-images.githubusercontent.com/22094846/173116451-18e758a0-3276-4a6c-900e-72eafa7f83de.png)

JaGeo replied:
This is the script that I am using for the workflow. I might need to try it in a completely new environment to make sure this is also not the reason for this problem.
[start_workflow.zip](https://github.com/materialsproject/atomate2/files/8880814/start_workflow.zip)

utf replied:
It seems like the workflow is using the same symprec throughout but it is quite high. I wonder if it will fix things if you use 1e-5?

JaGeo replied:
I will try! Thanks!

JaGeo replied:
You were correct. It was the symprec... I have now introduced a second symprec setting. 

I will try to further clean up the code in the next days and check the kpath setting.

JaGeo commented:
Open points: 
- [ ] Code is still a bit messy, needs some refactoring 
- [ ] Cleanup of documentation, making sure every important parameter can be somehow adapted
- [ ] Need to check kpath creation and correspondence with created cells again
- [ ] decision what we would like to save in the document (should some more fields live in a data store for large data?)
- [ ] write unit tests
- [ ] make tests on real systems with real supercell sizes
- [ ] potentially create another way to determine supercell matrices that prefers 90 degree angels

JaGeo commented:
@utf in case you have some time in the next weeks, I would be happy about Feedback on the document part.

I think we need to store force constants, born charges etc. to easily develop workflows for Grüneisen parameters/quasi-harmonic approximation.

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,329 @@
+    code: str,
+    displacement_data: dict[str, list],
+    total_energy: float,
+    epsilon_static: Matrix3D = None,
```
I wouldn't define all of these optional arguments here. Firstly, because it is duplication of default parameters and could result in the defaults getting out of sync with the document model. And secondly because if you want to add a new option to the document model you will have to remember to add it again here.

Instead just use  **kwargs and then change the function contents to something like:
```
    phonon_doc = PhononBSDOSDoc.from_forces_born(
         structure=structure,
         supercell_matrix=supercell_matrix,
         displacement=displacement,
         sym_reduce=sym_reduce,
         symprec=symprec,
         use_standard_primitive=use_standard_primitive,
         kpath_scheme=kpath_scheme,
         code=code,
         displacement_data=displacement_data,
         total_energy=total_energy,
         **kwargs
     )
```

JaGeo replied:
Thank you. I agree with this. I would however also add `born`, `epsilon_static` and `full_born` to the explicit keywords. They will be set in most of the cases.

utf replied:
Ok that sounds good. Although please see my comment above about full_born, I think we should be able to detect this automatically.

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,306 @@
+    def make(
+        self,
+        structure: Structure,
+        prev_vasp_dir: str | Path | None = None,
```
Instead of using `born_manual`, `epsilon_static_manual`, and `supercell_matrix` as dataclass fields, can you add `born`, `epsilon_static` and `supercell_matrix` to the make function arguments instead. And only calculate born etc if these are set to `None`.

utf commented:
> I think we need to store force constants, born charges etc. to easily develop workflows for Grüneisen parameters/quasi-harmonic approximation.

I think its a good idea to store the force constants too. Maybe there could be a flag to disable storing them if people are want to be very space conscious or e.g., if you are only doing convergence calculations.

JaGeo commented:
I have cleaned up the output document:
![image](https://user-images.githubusercontent.com/22094846/182648653-f28ffffa-8668-46d7-ad7f-c2af0c47a92c.png)

@utf , please let me know what you think about the new structure.

I will then proceed to fulfill the other open points on my todo list above. I will also provide a detailed check of the structure/kpath correspondence here as I think this is the only part where something else could go massively wrong. Then, I will think about useful unit tests.

JaGeo commented:
I am currently performing some more quantitative tests for ZnS. I personally think that the convergence parameters are quite strict (maybe too strict?) and I would not expect any major differences to the Abinit results for now. I will post the comparison here. 

Then, I will add more documentation and further clean up the code and I would ask you to review again.

JaGeo commented:
I have now performed a large-scale test with a supercell larger than 15 Angstrom in each direction for ZnS (mp-10695). I think it compares extremely well to the phonon band structure from Abinit and DFPT. 

**MP**
![image](https://user-images.githubusercontent.com/22094846/183833205-0e4a31e4-09f3-40d7-8788-ed970494ada3.png)

**This workflow**
![image](https://user-images.githubusercontent.com/22094846/183833265-0d219105-37c5-4412-a6f3-c90c4a31df18.png)

JaGeo commented:
I am now at the stage where I would need to write the unit tests and adapt requirements. I would be happy if another code review could be done by you @utf before so that I would not need to change the structure of the unit tests afterwards.

utf commented on src/atomate2/vasp/files.py:
```diff
@@ -177,7 +177,7 @@ def write_vasp_input_set(
         structure, prev_dir=prev_dir, potcar_spec=potcar_spec
     )
-
+    print(vis)
```
Remove unused code.

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,326 @@
+    displacement: float
+        displacement distance for phonons
+    min_length: float
+        min length of the supercell that will be build
```
```suggestion
        min length of the supercell that will be built
```

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,326 @@
+        with 3 90 degree angles
+    get_supercell_kwargs: additional arguments
+        to determine supercell
+    use_primitive_standard_structure: bool
+        this will enforce to start the phonon computation
+        from the primitive standard structure
+        according to Setyawan, W., & Curtarolo, S. (2010).
+        High-throughput electronic band structure calculations:
+        Challenges and tools. Computational Materials Science,
+        49(2), 299-312. doi:10.1016/j.commatsci.2010.05.010.
+        This makes it possible to use certain k-path definitions
+        with this workflow. Otherwise, we must rely on seekpath
+    use_conventional_standard_structure: bool
+        this will enforce to start the phonon computation
+        from the conventional standard structure
+        according to Setyawan, W., & Curtarolo, S. (2010).
+        High-throughput electronic band structure calculations:
+        Challenges and tools. Computational Materials Science,
+        49(2), 299-312. doi:10.1016/j.commatsci.2010.05.010.
+        We will however use seekpath and primitive structures
+        as determined by from phonopy to compute the phonon band structure
```
Maybe these should be combined into a single parameter "use_symmetrize_structure", which accepts "primitive", "conventional", or None.

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,326 @@
+        A maker to perform the computation of the DFT energy on the bulk.
+        Set to ``None`` to skip the
+        static energy computation
+    generate_phonon_diplsacements_kwargs: dict
```
```suggestion
    generate_phonon_displacements_kwargs: dict
```

JaGeo replied:
Will remove this and the next kwargs as they are not further used in the code.

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,326 @@
+        prev_vasp_dir: str | Path | None = None,
+        born_manual: List[Matrix3D] | None = None,
+        full_born: bool = True,
+        epsilon_static_manual: Matrix3D | None = None,
```
Can you remove the "manual" from epsilon and born kwargs?

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,326 @@
+            A pymatgen structure.
+        prev_vasp_dir : str or Path or None
+            A previous vasp calculation directory to use for copying outputs.
+        born_manual: Matrix3D
```
Is it possible to remove the `full_born` parameter and just figure out with the full or reduced born has been passed by checking the length of the born array. E.g., using some pseudocode like:

```python
if len(born) == len(structure):
    full_born = False
```

This will have to be put inside the job where the born is actually used as the structure might not be available otherwise.

Also, if you go down this route, you can specify that either the reduced or full born description can be supplied, in the born parameter docstring.

JaGeo replied:
Thank for the suggestion.

JaGeo replied:
Will add the following and test if it works: 
`if len(born) == len(structure):
    full_born = True`

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,326 @@
+            structure = conv_job.output
+
+        # if supercell_matrix is None, supercell size will be determined
+        if supercell_matrix is not None:
+            self.supercell_matrix = None
+        else:
+            supercell_job = get_supercell_size(
+                structure,
+                self.min_length,
+                self.prefer_90_degrees,
+                **self.get_supercell_size_kwargs,
+            )
+            jobs.append(supercell_job)
+            self.supercell_matrix = supercell_job.output
```
I think we shouldn't set `self.supercell_matrix` and instead this could just be:
```suggestion
        if supercell_matrix is None:
            supercell_job = get_supercell_size(
                structure,
                self.min_length,
                self.prefer_90_degrees,
                **self.get_supercell_size_kwargs,
            )
            jobs.append(supercell_job)
            supercell_matrix = supercell_job.output
```

And later on in this function replace `self.supercell_matrix` with just `supercell_matrix`

JaGeo replied:
I agree and it would also fix the bug that I had overlooked so far ...

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,326 @@
+        jobs.append(vasp_displacement_calcs)
+
+        # Computation of BORN charges
+        if self.born_maker is not None and (
+            born_manual is None or epsilon_static_manual is None
+        ):
+            born_job = self.born_maker.make(structure)
+            jobs.append(born_job)
```
Can you update this to subsume the additional born logic later on in the function. Note this also assumes you rename born_manual and epsilon_static_manual, and are able to remove the full_born parameter:


```suggestion
        if self.born_maker is not None and (
            born_manual is None or epsilon_static_manual is None
        ):
            born_job = self.born_maker.make(structure)
            jobs.append(born_job)
			
            # I am not happy how we currently access "born" charges
            # This is very vasp specific code
            epsilon_static = born_job.output.calcs_reversed[0].output.epsilon_static
            born = born_job.output.calcs_reversed[0].output.outcar["born"]
            born_run_job_dir = born_job.output.dir_name
            born_run_uuid = born_job.output.uuid
        else:
            born_run_job_dir = None
            born_run_uuid = None
```

JaGeo replied:
I updated it. Moved the static run after this whole born logic as well.

utf commented on src/atomate2/vasp/schemas/phonons.py:
```diff
@@ -0,0 +1,474 @@
+        description="Structure of Materials Project.",
+    )
+
+    ph_bs: PhononBandStructureSymmLine = Field(
```
Can you rename just for clarity. E.g., to `phonon_bandstructure`

JaGeo replied:
Done

utf commented on src/atomate2/vasp/schemas/phonons.py:
```diff
@@ -0,0 +1,474 @@
+        description="Phonon band structure object.",
+    )
+
+    ph_dos: PhononDos = Field(
```
And similarly to `phonon_dos`.

JaGeo replied:
Done

utf commented on src/atomate2/vasp/schemas/phonons.py:
```diff
@@ -0,0 +1,474 @@
+        description="Phonon density of states object.",
+    )
+
+    free_energy_list: List[float] = Field(
```
You can remove the list, so it is just `free_energy` or `free_energies`

JaGeo replied:
Done

utf commented on src/atomate2/vasp/schemas/phonons.py:
```diff
@@ -0,0 +1,474 @@
+    epsilon_static: Matrix3D = Field(
+        None, description="The high-frequency dielectric constant"
+    )
+    dft_phononwebsite: dict = Field(
```
Do you have an idea of the size of this data? Is it very big?

JaGeo replied:
Will check this! Thanks for the hint!

JaGeo replied:
I decided to simply write a file to the last folder including this information so that it does not enlarge the database. Currently, this is only a dict and it has to be of the same size as the bandstructure objects.

utf reviewed (commented):
Hi @JaGeo, apologies for the delay looking at this PR, I was on annual leave and I then I just started a new job this week.

This is looking in great shape. I left some additional comments. It is basically tidying at this point, with a few renamings. Once those are done I think the last thing are the tests.

Thanks again!

JaGeo commented:
Thanks for the review! And, no worries :) . I can simply switch to another task if something needs more time. Congrats again on the new position!  
I will try to implement the changes in the coming weeks and then start with the unit tests. I will also try to use it for some projects soon to make sure it is also tested on a large number of different structures. @ab5424 tested the initial implementation on many structures but we changed many things due to relying on atomate2 defaults.

JaGeo commented:
I will start with the tests during this week. I currently, however, have problems with other tests when switching to a newer version of pymatgen. I guess this might be partially solved when the new defect workflow (https://github.com/materialsproject/atomate2/pull/133) is merged.

utf commented on pyproject.toml:
```diff
@@ -55,7 +55,8 @@ tests = [
     "pydantic==1.9.0",
-    "pymatgen==2022.7.8",
+    "pymatgen==2022.8.23",
+    "pymatgen-analysis-defects==2022.8.24",
```
Please remove this! This will be an option dependency.

JaGeo replied:
I will.

JaGeo replied:
I updated the non-strict installation part as well. I hope this is fine.

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,318 @@
+    get_supercell_kwargs: additional arguments
+        to determine supercell
+    use_symmetrized_structure: str
+        allowed strings: "primitive", "conventional", None
+
+        "primitive" will enforce to start the phonon computation
+        from the primitive standard structure
+        according to Setyawan, W., & Curtarolo, S. (2010).
+        High-throughput electronic band structure calculations:
+        Challenges and tools. Computational Materials Science,
+        49(2), 299-312. doi:10.1016/j.commatsci.2010.05.010.
+        This makes it possible to use certain k-path definitions
+        with this workflow. Otherwise, we must rely on seekpath
+
+        "conventional" will enforce to start the phonon computation
+        from the conventional standard structure
+        according to Setyawan, W., & Curtarolo, S. (2010).
+        High-throughput electronic band structure calculations:
+        Challenges and tools. Computational Materials Science,
+        49(2), 299-312. doi:10.1016/j.commatsci.2010.05.010.
+        We will however use seekpath and primitive structures
+        as determined by from phonopy to compute the phonon band structure
```
Use rst list formatting syntax instead:
```suggestion
        allowed strings: "primitive", "conventional", None
        
        - "primitive" will enforce to start the phonon computation
          from the primitive standard structure
          according to Setyawan, W., & Curtarolo, S. (2010).
          High-throughput electronic band structure calculations:
          Challenges and tools. Computational Materials Science,
          49(2), 299-312. doi:10.1016/j.commatsci.2010.05.010.
          This makes it possible to use certain k-path definitions
          with this workflow. Otherwise, we must rely on seekpath
        - "conventional" will enforce to start the phonon computation
          from the conventional standard structure
          according to Setyawan, W., & Curtarolo, S. (2010).
          High-throughput electronic band structure calculations:
          Challenges and tools. Computational Materials Science,
          49(2), 299-312. doi:10.1016/j.commatsci.2010.05.010.
          We will however use seekpath and primitive structures
          as determined by from phonopy to compute the phonon band structure
```

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,318 @@
+        if self.static_energy_maker is not None and total_dft_energy is None:
+            static_job = self.static_energy_maker.make(structure=structure)
+            jobs.append(static_job)
+            total_energy = static_job.output.output.energy
```
```suggestion
            total_dft_energy = static_job.output.output.energy
```

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,318 @@
+            static_run_job_dir = static_job.output.dir_name
+            static_run_uuid = static_job.output.uuid
+        else:
+            total_energy = total_dft_energy
```
And then you can remove this line

JaGeo replied:
Thanks. Will also change all other side effects of this change. I should probably run some of the automatic systems already now to spot such things. I wanted to wait until the unit tests are implemented as I always suspect potential problems.

utf commented on src/atomate2/vasp/flows/phonons.py:
```diff
@@ -0,0 +1,318 @@
+            born_run_job_dir = born_job.output.dir_name
+            born_run_uuid = born_job.output.uuid
+        else:
+            epsilon_static = epsilon_static
+            born = born
```
Remove these lines

JaGeo replied:
In Germany, we call this "betriebsblind". Thanks.

utf replied:
That's a great word!

utf commented on src/atomate2/vasp/jobs/base.py:
```diff
@@ -133,7 +133,6 @@ def make(self, structure: Structure, prev_vasp_dir: str | Path | None = None):
 
         if "from_prev" not in self.write_input_set_kwargs:
             self.write_input_set_kwargs["from_prev"] = from_prev
-
```
Can you add this back in?

JaGeo replied:
Yes, of course.

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,447 @@
+    npoints_band: int = 100,
+    kpoint_density_dos: int = 7000,
+    tol_imaginary_modes: float = 1e-5,
+    tmin=0,
```
Please see other comment about moving these optional arguments into **kwargs.

JaGeo replied:
Yes, will do. I am just sometimes a bit worried that people don't see through these **kwargs as it complicates usage in my opinion. Not sure everyone immediately understands the Structure.from_file(...) in pymatgen for example.

utf replied:
Yes, I also have this worry. Many people do not understand it. Perhaps you could add an explanation in the docstring pointing people to this other method?

Also, as this is a common practice in atomate2, we could even have a dedicated page in the docs about how to use atomate2 effectively, with some common design principles.

JaGeo replied:
Both suggestions sound like a great idea. 

I was also thinking about whether there should be a GUI at some point for people to build larger workflows by drag and drop or similar. You might have thought about it already and it would also be a massive amount of work ... but it's really hard to get into the usage workflows at the moment. Or we add even more tutorials...

utf replied:
I think in the short term more tutorials is definitely needed. I was even thinking about adding a tutorial for most substantial workflows. E.g., we could add a workflow on how to do phonons using atomate2. It is more work from the developer side but it would definitely increase adoption!

As for a GUI, I know @mkhorton has a great vision for this, and its also something I'd really want to do. I'm not sure how to make it happen, specially regarding funding for the project. Realistically, it will need an experienced web developer working for ~1 year minimum. 

I actually think this could be something which a spin-off company would be well suited for. Create a web application that could be licensed out to industry to fund the development (academia could be free/get a discount). But I'm aware this doesn't fit into the ethos of MP.

JaGeo replied:
I agree that this will be challenging. Maybe, we could also think about a template for developing your own workflow and mention the common steps in workflow development with important classes/methods for this? This is how I would start implementing something like this with a new co-worker as well. I don't think that we can expect typical chemists/physicits to navigate new and complex codes very well. 

I am really happy about the test documentation.

JaGeo replied:
@davidwaroquiers I would love to join the discussion as it actually fits one of my broader and long-term interests.

utf commented on src/atomate2/vasp/schemas/phonons.py:
```diff
@@ -0,0 +1,469 @@
+        """
+            get high-symmetry points in k-space
+            Parameters
+        -   ---------
```
Weird formatting?

Also there should be a blank line between the first line and the Parameters line

JaGeo replied:
Uups.

utf reviewed (commented):
Hi @JaGeo. I fixed the issue with pymatgen, so if you pull in the changes on main that will hopefully fix your problem. I also left some final comments which would be useful to address before doing any testing.

Thanks!

JaGeo commented:
Thanks! Will further refine it.

JaGeo commented:
To test all potential paths through this workflow, I will add 4 more sets of test data. Unfortunately, this is necessary due to the options for kpath settings and cell definitions. Also, I need to test a compound different from Si to actually have useful BORN charges.

JaGeo commented:
I have added the last test. I found some issues in the normalization of the free energy for non-primitive cells that I have corrected and I have decided to change the input from total_dft_energy_per_cell to total_dft_energy_per_formula unit so that different cell definitions should be easier to handle.

I will probably now go through the typical workflow checks and then we could re-run workflows here.

If someone else would like to test this workflow even more, I would be happy about it. From my side, it will take a few weeks to run several (10-20) MP structures and compare to the results from DFPT and Abinit. Band structure of Si at least agrees with the MP within 2 cm-1.

JaGeo commented:
Would be ready for another round of review and the automatic tests from side at least.

JaGeo commented:
Thanks. Stupid mistake from my side...

phonopy is okay as an requirement?

utf commented:
Please could you add it as an optional dependency for now? And then feel free to edit the GitHub actions to also install that optional dependency set.

JaGeo commented:
Thanks. Yes, I think the latter is the issue. It's already an optional requirement.

JaGeo commented:
Sorry... Added it to the wrong workflow ....

JaGeo commented:
Had to add seekpath to the optional requirements [phonons] as well. Everytime I install phonopy, I forget to install seekpath ...

JaGeo commented:
Tests have passed. I would reply to any further comments later today or next week.

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,398 @@
+
+    Parameters
+    ----------
+    structure: Structure object
```
Same issue with missing description.

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,398 @@
+    born: Matrix3D
+        Born charges
+    kwargs:
+        additional arguments that are passed
```
Say where they are to `PhononBSDOSDoc.from_forces_born`.

utf commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,398 @@
+
+    Parameters
+    ----------
+    displacements
+    structure: original structure for meta data
+    supercell_matrix: supercell matrix for meta data
```
Missing types and descriptions.

utf reviewed (commented):
Hi @JaGeo, a few more observations.

I'm really happy how this PR turned out. Thanks for all your edits. Personally, I think once you've made these changes we can merge this and then do the more substantial testing afterwards.

JaGeo commented:
Thanks. Will go through all of the suggested changes today! 
I am currently trying to setup everything on another computer so that I can hopefully test more compounds faster. I still sometimes have weird issues with custodian and VASP for extremely simple systems that I need to resolve ...

JaGeo commented on src/atomate2/vasp/jobs/phonons.py:
```diff
@@ -0,0 +1,398 @@
+
+
+@job
+def structure_to_primitive(structure: Structure, symprec: float):
```
I think it is probably fine to use the defaults here. Will move the functions, rename and change to default symprec.

JaGeo commented:
I hope I added all suggestions. I made one more change on the tests. I realized on the weekend that I had several useless statements in there ... 

We could rerun the tests again. I hope that I changed the installation configurations correctly. I will be really happy if you don't need to restart the automatic tests by hand every time I make a mistake ...

JaGeo commented:
Yeah, of course, more typos. I usually use the read-out-loud function everywhere to find such things... Now, I think it should be fine ;).

utf commented on src/atomate2/common/jobs.py:
```diff
@@ -6,25 +6,46 @@
-        A structure.
-    symprec : float
-        The symmetry precision.
+        structure: Structure object
```
I don't think these should be indented an extra level.

JaGeo replied:
Corrected.

JaGeo replied:
And, tests have to be restarted, of course. Sorry.

utf commented:
>  I will be really happy if you don't need to restart the automatic tests by hand every time I make a mistake ...

I think once this is merged GitHub won't require me to do it manually any more, since you will officially be a contributors. 🤞 

One final comment. Are you happy if I merge once that is resolved?

JaGeo commented:
Yes, I would be happy if this is merged. I will then run several test structures, use it in my research and start a new pull-request to fix potential bugs in case they show up.

utf commented:
Thanks very much @JaGeo. This is an exemplar workflow!

JaGeo commented:
Thank you so much
